### PR TITLE
UIInteraction should be added to visibility propagation view

### PR DESCRIPTION
**Pull request by pvollan**
#### bea78b24700883adae93a0992eb3005d4d362891
<pre>
UIInteraction should be added to visibility propagation view
<a href="https://bugs.webkit.org/show_bug.cgi?id=266275">https://bugs.webkit.org/show_bug.cgi?id=266275</a>
<a href="https://rdar.apple.com/119434342">rdar://119434342</a>

Reviewed by Brent Fulgham.

UIInteraction should be added to visibility propagation view, not the Web view itself.

* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _setupVisibilityPropagationViewForWebProcess]):
(-[WKContentView _setupVisibilityPropagationViewForGPUProcess]):
(-[WKContentView _removeVisibilityPropagationViewForWebProcess]):
(-[WKContentView _removeVisibilityPropagationViewForGPUProcess]):

Canonical link: <a href="https://commits.webkit.org/271945@main">https://commits.webkit.org/271945@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/b6c73ec224eeb328e8ce8194cb5369b0dc72a3f6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30084 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8748 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31397 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32586 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27198 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10946 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5995 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27221 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30383 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7336 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25650 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6264 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6421 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33922 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27429 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27175 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32601 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6363 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4536 "Found 1 new test failure: fullscreen/full-screen-layer-dump.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30401 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8113 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7117 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3887 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6891 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/UIProcess/ios/WKContentView.mm [hunk 1]
```diff
@@ -173,7 +173,9 @@ @implementation WKContentView {
 #if USE(EXTENSIONKIT)
     RetainPtr<id<UIInteraction>> _visibilityPropagationInteractionForWebProcess;
     RetainPtr<id<UIInteraction>> _visibilityPropagationInteractionForGPUProcess;
-#endif
+    RetainPtr<UIView> _visibilityPropagationViewForWebProcess;
+    RetainPtr<UIView> _visibilityPropagationViewForGPUProcess;
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/ios/WKContentView.mm]: I'm afraid I don't understand why we need to create a separate set of `UIView`s here. Are these meant to represent UI Interactions that should be relayed to the Web and GPU Process? Why are they distinct from the `UIView` that underlies the `WKContentView` object itself?

Comment (pvollan) [reply Source/WebKit/UIProcess/ios/WKContentView.mm]: This is not strictly necessary, it was just to make clear that we are creating UIViews in the extension case, and not the specialized sub-classes.

Thanks for reviewing!

File: Source/WebKit/UIProcess/ios/WKContentView.mm [hunk 2]
```diff
@@ -274,12 +277,16 @@ - (instancetype)_commonInitializationWithProcessPool:(WebKit::WebProcessPool&)pr
...
 #if USE(EXTENSIONKIT)
-    if (!_visibilityPropagationInteractionForWebProcess) {
+    if (!_visibilityPropagationInteractionForWebProcess && !_visibilityPropagationViewForWebProcess) {
         SEL selector = NSSelectorFromString(@"createVisibilityPropagationInteraction");
         if ([_page->process().extensionProcess().get() respondsToSelector:selector])
             _visibilityPropagationInteractionForWebProcess = [_page->process().extensionProcess() performSelector:selector];
-        if (_visibilityPropagationInteractionForWebProcess)
-            [self addInteraction:_visibilityPropagationInteractionForWebProcess.get()];
+        if (_visibilityPropagationInteractionForWebProcess) {
+            _visibilityPropagationViewForWebProcess = adoptNS([[UIView alloc] initWithFrame:CGRectZero]);
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/ios/WKContentView.mm]: So these are "invisible" UIViews that the user cannot interact with, it seems.

File: Source/WebKit/UIProcess/ios/WKContentView.mm [hunk 3]
```diff
@@ -338,7 +349,7 @@ - (void)_removeVisibilityPropagationViewForWebProcess
 {
 #if USE(EXTENSIONKIT)
     if (_visibilityPropagationInteractionForWebProcess) {
-        [self removeInteraction:_visibilityPropagationInteractionForWebProcess.get()];
+        [_visibilityPropagationViewForWebProcess removeInteraction:_visibilityPropagationInteractionForWebProcess.get()];
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/ios/WKContentView.mm]: Do we need to remove the view from the parent, too?

**Review by brentfulgham — APPROVED** [review]
Looks fine, but I think we could get away with fewer new UIViews in the final implementation.

File: Source/WebKit/UIProcess/ios/WKContentView.mm (hunk 3 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/ios/WKContentView.mm]: Good point! That is done in the code below.

Thanks for reviewing!

File: Source/WebKit/UIProcess/ios/WKContentView.mm (hunk 2 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/ios/WKContentView.mm]: Exactly, their only purpose is to take care of the visibility propagation.

**Pull request merged**